File: src/main/scala/com/kotori316/fluidtank/recipes/CombineRecipe.java
package com.kotori316.fluidtank.recipes;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import com.google.gson.JsonObject;
import net.minecraft.core.NonNullList;
import net.minecraft.data.recipes.FinishedRecipe;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.inventory.CraftingContainer;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.CraftingBookCategory;
import net.minecraft.world.item.crafting.CustomRecipe;
import net.minecraft.world.item.crafting.Ingredient;
import net.minecraft.world.item.crafting.RecipeSerializer;
import net.minecraft.world.level.Level;
import org.apache.commons.lang3.tuple.Pair;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.Nullable;
import scala.jdk.javaapi.StreamConverters;

import com.kotori316.fluidtank.ModObjects;
import com.kotori316.fluidtank.Utils;
import com.kotori316.fluidtank.blocks.BlockTank;
import com.kotori316.fluidtank.fluids.FluidAction;
import com.kotori316.fluidtank.fluids.FluidAmount;
import com.kotori316.fluidtank.fluids.FluidKey;
import com.kotori316.fluidtank.items.ItemBlockTank;
import com.kotori316.fluidtank.items.TankItemFluidHandler;

public class CombineRecipe extends CustomRecipe {
    private static final Logger LOGGER = Utils.getLogger(CombineRecipe.class);
    public static final RecipeSerializer<CombineRecipe> SERIALIZER = new Serializer();
    public static final String LOCATION = "fluidtank:combine_tanks";
    private final Ingredient tanks;

    public CombineRecipe(ResourceLocation location, Ingredient tanks) {
        super(location, CraftingBookCategory.MISC);
        this.tanks = tanks;
        LOGGER.debug("{} instance({}) created", getClass().getSimpleName(), location);
    }

    @Override
    public boolean matches(CraftingContainer inv, @Nullable Level worldIn) {
        // Check all items are tanks.
        boolean allTanks = IntStream.range(0, inv.getContainerSize())
            .mapToObj(inv::getItem).filter(s -> !s.isEmpty()).allMatch(tanks);
        if (!allTanks) return false;
        long tankCount = IntStream.range(0, inv.getContainerSize())
            .mapToObj(inv::getItem)
            .filter(s -> !s.isEmpty()).filter(tanks).count();
        if (tankCount < 2) return false;
        // Check all tanks have the same fluid.
        List<FluidAmount> fluids = IntStream.range(0, inv.getContainerSize())
            .mapToObj(inv::getItem)
            .map(s -> getHandler(s)
                .map(TankItemFluidHandler::getFluid)
                .orElse(FluidAmount.EMPTY()))
            .toList();
        boolean allSameFluid = fluids.stream()
                                   .map(FluidKey::from)
                                   .filter(FluidKey::isDefined)
                                   .distinct().count() == 1;
        if (!allSameFluid) return false;
        long totalAmount = fluids.stream().mapToLong(FluidAmount::amount).sum();
        return getMaxCapacityTank(inv).map(p -> p.getRight() >= totalAmount).orElse(false);
    }

    private static Optional<Pair<ItemStack, Long>> getMaxCapacityTank(CraftingContainer inv) {
        return getMaxCapacityTank(IntStream.range(0, inv.getContainerSize())
            .mapToObj(inv::getItem));
    }

    static Optional<Pair<ItemStack, Long>> getMaxCapacityTank(Stream<ItemStack> stream) {
        return stream
            .filter(s -> !s.isEmpty())
            .flatMap(s -> getHandler(s).stream())
            .map(h -> Pair.of(h.getContainer(), h.getCapacity()))
            .max(Comparator.comparing(Pair::getRight))
            .map(p -> Pair.of(p.getLeft().copy(), p.getRight()));
    }

    @Override
    public ItemStack assemble(CraftingContainer inv) {
        Optional<FluidAmount> fluid = IntStream.range(0, inv.getContainerSize())
            .mapToObj(inv::getItem)
            .map(s -> getHandler(s)
                .map(TankItemFluidHandler::getFluid)
                .orElse(FluidAmount.EMPTY()))
            .filter(FluidAmount::nonEmpty)
            .reduce(FluidAmount::$plus);
        Optional<ItemStack> tank = getMaxCapacityTank(inv)
            .flatMap(p -> getHandler(copyStackWithSize(p.getLeft(), 1))
                .flatMap(h ->
                    fluid.map(f -> {
                        h.drain(h.getFluidInTank(0), FluidAction.EXECUTE);
                        h.fill(f, FluidAction.EXECUTE);
                        return h.getContainer();
                    })));
        return tank.orElse(ItemStack.EMPTY);
    }

    @Override
    public NonNullList<ItemStack> getRemainingItems(CraftingContainer inv) {
        Optional<ItemStack> stack = getMaxCapacityTank(inv).map(Pair::getLeft);

        NonNullList<ItemStack> nn = NonNullList.withSize(inv.getContainerSize(), ItemStack.EMPTY);
        for (int i = 0; i < nn.size(); i++) {
            ItemStack item = inv.getItem(i);
            if (stack.filter(s -> s.sameItem(item)).isPresent()) {
                stack = Optional.empty();
            } else {
                ItemStack leave = getHandler(copyStackWithSize(item, 1))
                    .map(h -> {
                        h.drain(h.getFluidInTank(0), FluidAction.EXECUTE);
                        return h.getContainer();
                    }).orElse(item);
                nn.set(i, leave);
            }
        }
        return nn;
    }

    @Override
    public boolean canCraftInDimensions(int width, int height) {
        return width * height >= 2;
    }

    @Override
    public RecipeSerializer<?> getSerializer() {
        return SERIALIZER;
    }

    private static Optional<TankItemFluidHandler> getHandler(ItemStack stack) {
        if (stack.getItem() instanceof ItemBlockTank tankItem) {
            return Optional.of(new TankItemFluidHandler(tankItem.blockTank().tier(), stack));
        } else {
            return Optional.empty();
        }
    }

    private static ItemStack copyStackWithSize(ItemStack stack, int size) {
        var copied = stack.copy();
        copied.setCount(size);
        return copied;
    }

    private static final class Serializer implements RecipeSerializer<CombineRecipe> {
        @Override
        public CombineRecipe fromJson(ResourceLocation pRecipeId, JsonObject pSerializedRecipe) {
            var tanks = tankList();
            return new CombineRecipe(pRecipeId, tanks);
        }

        @Override
        public CombineRecipe fromNetwork(ResourceLocation pRecipeId, FriendlyByteBuf pBuffer) {
            var tanks = Ingredient.fromNetwork(pBuffer);
            return new CombineRecipe(pRecipeId, tanks);
        }

        @Override
        public void toNetwork(FriendlyByteBuf pBuffer, CombineRecipe recipe) {
            recipe.tanks.toNetwork(pBuffer);
        }

        private static Ingredient tankList() {
            Stream<BlockTank> tankStream = StreamConverters.asJavaSeqStream(ModObjects.blockTanks());
            Predicate<BlockTank> filter = t -> t.tier().isNormalTier();
            return Ingredient.of(tankStream.filter(filter).map(ItemStack::new));
        }
    }

    public record CombineFinishedRecipe(ResourceLocation recipeId) implements FinishedRecipe {
        @Override
        public void serializeRecipeData(JsonObject pJson) {

        }

        @Override
        public ResourceLocation getId() {
            return recipeId;
        }

        @Override
        public RecipeSerializer<?> getType() {
            return SERIALIZER;
        }

        @Nullable
        @Override
        public JsonObject serializeAdvancement() {
            return null;
        }

        @Override
        public ResourceLocation getAdvancementId() {
            return new ResourceLocation("");
        }
    }
}

File: src/main/scala/com/kotori316/fluidtank/recipes/CombineRecipe.scala
package com.kotori316.fluidtank.recipes

import cats.syntax.foldable._
import com.kotori316.fluidtank.blocks.BlockTank
import com.kotori316.fluidtank.fluids.FluidKey
import com.kotori316.fluidtank.items.{ItemBlockTank, TankItemFluidHandler}
import com.kotori316.fluidtank.{Config, ModObjects, Utils}
import net.minecraft.core.NonNullList
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.inventory.CraftingContainer
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.crafting.{CustomRecipe, Ingredient, RecipeSerializer, SimpleRecipeSerializer}
import net.minecraft.world.level.Level
import net.minecraftforge.fluids.capability.IFluidHandler
import net.minecraftforge.items.ItemHandlerHelper

final class CombineRecipe(location: ResourceLocation) extends CustomRecipe(location) {

  import CombineRecipe._

  LOGGER.debug("Recipe instance of ConvertInvisibleRecipe({}) was created.", location)

  override def canCraftInDimensions(width: Int, height: Int): Boolean = width * height >= 2

  override def getSerializer: RecipeSerializer[_] = SERIALIZER

  override def matches(inv: CraftingContainer, worldIn: Level): Boolean = {
    val tanks = tankList()
    val isAllItemTank = (0 until inv.getContainerSize)
      .map(inv.getItem)
      .filterNot(_.isEmpty)
      .forall(tanks.test)
    if (!isAllItemTank) return false
    val tankCount = (0 until inv.getContainerSize)
      .map(inv.getItem)
      .filterNot(_.isEmpty)
      .count(tanks.test)
    if (tankCount < 2) return false
    // Check all tanks have the same fluid.
    val fluids = (0 until inv.getContainerSize)
      .map(inv.getItem)
      .flatMap(s => getHandler(s).map(_.getFluid))
    val allSameFluid = fluids
      .map(FluidKey.from)
      .filter(_.isDefined)
      .distinct.lengthIs == 1
    if (!allSameFluid) return false
    val totalAmount = fluids.map(_.amount).sum
    getMaxCapacityTank(inv).exists { case (_, capacity) => capacity >= totalAmount }
  }

  override def assemble(inv: CraftingContainer): ItemStack = {
    val fluid = (0 until inv.getContainerSize).toList
      .map(inv.getItem)
      .flatMap(s => getHandler(s).map(_.getFluid))
      .filter(_.nonEmpty)
      .combineAll
    val tank = for {
      (tankItem, _) <- getMaxCapacityTank(inv)
      handler <- getHandler(ItemHandlerHelper.copyStackWithSize(tankItem, 1))
    } yield {
      handler.drain(Int.MaxValue, IFluidHandler.FluidAction.EXECUTE)
      handler.fill(fluid.toStack, IFluidHandler.FluidAction.EXECUTE)
      handler.getContainer
    }
    tank.getOrElse(ItemStack.EMPTY)
  }

  override def getRemainingItems(inv: CraftingContainer): NonNullList[ItemStack] = {
    var maxTank: ItemStack = getMaxCapacityTank(inv).map(_._1).getOrElse(ItemStack.EMPTY)
    val nn = NonNullList.withSize(inv.getContainerSize, ItemStack.EMPTY)
    for (i <- 0 until nn.size()) {
      val item = inv.getItem(i)
      if (item sameItem maxTank) {
        maxTank = ItemStack.EMPTY
      } else {
        val leave = getHandler(ItemHandlerHelper.copyStackWithSize(item, 1))
          .map { h => h.drain(Int.MaxValue, IFluidHandler.FluidAction.EXECUTE); h.getContainer }
          .getOrElse(item)
        nn.set(i, leave)
      }
    }
    nn
  }

  private def tankList(): Ingredient = {
    val filter = if (Config.content.usableUnavailableTankInRecipe.get()) {
      (t: BlockTank) => t.tier.isNormalTier
    } else {
      (t: BlockTank) => t.tier.isNormalTier && t.tier.hasWayToCreate
    }
    Ingredient.of(ModObjects.blockTanks.filter(filter).map(new ItemStack(_)): _*)
  }

  private def getMaxCapacityTank(inv: CraftingContainer): Option[(ItemStack, Long)] = {
    getMaxCapacityTank((0 until inv.getContainerSize).map(inv.getItem))
  }

  private def getMaxCapacityTank(stacks: Seq[ItemStack]): Option[(ItemStack, Long)] = {
    stacks
      .filterNot(_.isEmpty)
      .flatMap(getHandler)
      .maxByOption(_.getCapacity)
      .map(f => (f.getContainer.copy(), f.getCapacity))
  }

  private def getHandler(stack: ItemStack): Option[TankItemFluidHandler] = {
    stack.getItem match {
      case tankItem: ItemBlockTank => Option(new TankItemFluidHandler(tankItem.blockTank.tier, stack))
      case _ => None
    }
  }
}

object CombineRecipe {
  private final val LOGGER = Utils.getLogger(classOf[CombineRecipe])
  final val SERIALIZER = new SimpleRecipeSerializer(new CombineRecipe(_))
  final val LOCATION = "fluidtank:combine_tanks"
}
